Adds a 'vultr_reverse_name' resource

cimnine opened the pull request:
This PR implements a new resource for [setting reverse dns entries][vultr_rdns].

[vultr_rdns]: https://www.vultr.com/docs/how-to-create-reverse-dns-or-ptr-records-in-the-vultr-control-panel

An extra resource has been chosen, because in order to set the reverse name, the IP(s) of the instance have to be known. And  that's only the case after the instance has been created.

The syntax is like this:

```terraform
// Define reverse name for main IPv4
resource "vultr_reverse_name" "example_v4" {
  instance_id = "${vultr_instance.example.id}"
  ip          = "${vultr_instance.example.ipv4_address}"
  name        = "example.com"
}

// Define reverse name for main IPv6
resource "vultr_reverse_name" "example_v6" {
  instance_id = "${vultr_instance.example.id}"
  ip          = "${vultr_instance.example.ipv6_addresses[0]}"
  name        = "example.com"
}
```

As you see, it works for IPv4 and IPv6 likewise. (Although internally they are handled by two different API endpoints. But the behavior as observed by the is very similar. That's why I've decided to combine the IPv4 and IPv6 case into one resource type.)

When destroying an IPv4 reverse name resource, nothing actually happens. That's because there is no API endpoint by Vultr to "restore the default name".

I think this resolves #34 and #56.

cimnine committed:
Implements 'vultr_reverse_name' resource

It works for IPv4 and IPv6 likewise.
An extra resource has been chosen, because in order to set the
reverse name, the IP(s) of the instance have to be known. And
that's only after they have been created.

When an IPv4 reverse name is deleted, nothing really happens.
This is because there is not API of Vultr to 'reset' the reverse
name.

cimnine committed:
Fix typo in README.md

squat commented on vultr/resource_reverse_name.go:
```diff
@@ -0,0 +1,182 @@
+func resourceReverseNameDelete(d *schema.ResourceData, meta interface{}) error {
+	id, ip, err := splitID(d.Id())
+
+	log.Printf("[INFO] Destroying resource reverse name for ip (%s)", ip)
```
I think “resource” here is unneeded.
Also, `IP` should be capitalized.

squat commented on vultr/resource_reverse_name.go:
```diff
@@ -0,0 +1,182 @@
+
+	if isIPv4(ip) {
+		// The reverse name can not be "unset", so doing nothing here
+
```
This newline is extra

squat commented on vultr/resource_reverse_name.go:
```diff
@@ -0,0 +1,182 @@
+		}
+	}
+
+	log.Printf("Can't get reverse name of IPv6 (%s) on instance (%s). Assuming that this IPv6 was deleted.", ip, id)
```
In general this provider prefixes log statements with the log level. Can you add that here?

squat commented on vultr/resource_reverse_name.go:
```diff
@@ -0,0 +1,182 @@
+			}
+		}
+
+		log.Printf("Can't get reverse name of IPv4 (%s) on instance (%s). Assuming that this IPv4 was deleted.", ip, id)
```
Also add prefix here.

squat commented on vultr/resource_reverse_name.go:
```diff
@@ -0,0 +1,182 @@
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
```
It looks like `k` and `d` are unused in this func. Can you replace the parameter names with `_` so it’s clearer?

squat commented:
@cimnine thank you for making the PR! I made a few notes for small changes and I have one question about the `name` attribute. Otherwise this looks great!

cimnine commented on vultr/resource_reverse_name.go:
```diff
@@ -0,0 +1,182 @@
+
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
```
It's optional in the sense that there is a default value on the Vultr side.

squat replied:
But that’s the option between creating the resource or not creating it, not the option between supplying  name or not, no?

cimnine replied:
I've changed this to be required and introduced a `vultr_reverse_name` _data source_ instead, which allows to query reverse names that are not declared / managed in terraform.

squat replied:
Thanks! I think that’s the correct approach.

cimnine committed:
Fix style

cimnine commented:
Thank you for your feedback!

I've implemented all your suggestions and added some additional validation.

Because adding additional validation required two external dependencies, and glide updates all dependencies when adding new ones, this PR suddenly got huge :-/

Maybe it's time to move on and `go mod init github.com/squat/terraform-provider-vultr` :wink:

squat commented:
Hmm I don’t believe that adding the validation dependency really needs 9,000,000 new lines of code, not diff’d lines but added lines. Can you please try `make vendor` to remove the unneeded files?

cimnine committed:
implement validation

cimnine committed:
add dependencies to glide

Command used:

```bash
glide get \
  --quick \
  -v \
  "golang.org/x/net/idna" \
  "github.com/hashicorp/terraform/helper/validation"
```

cimnine committed:
Change the 'name' to required

Even if it's technically speaking not required, this resource
doesn't make sense if isnt'.

cimnine committed:
Implemented a 'vultr_reverse_name' data source

For querying existing reverse names without declaring them
through terraform.

cimnine committed:
Add all vendor

cimnine commented:
Unfortunately it's not possible to pass the build and have less code changes in vendor. (Or I just don't get how.) Maybe you could give it a try?

We could argue if it's really necessary to add these dependencies. The IP checker isn't, that's trivial to implement. But the conversion between IDN and ASCII domain name representations is not easy to implement. And I believe that conversion is necessary in order to reliably compare whether domain names have changed.

squat commented:
I’ll take a look at the vendoring commit

cimnine commented:
As you might have seen I've went ahead (maybe too far?) and presented a way that creates less changes to the code base by migrating over to the `go mod` ecosystem.

Feel free to close these PRs anytime should you not be comfortable with the way I was approaching things.

~Chris
